src/HexagonOffload.cpp

```diff
@@ -0,0 +1,322 @@
+        Buffer code(type_of<uint8_t>(), {(int)size}, nullptr, name);
+        memcpy(code.host_ptr(), buffer, (int)size);
+
+        Expr ptr_0 = Load::make(type_of<uint8_t>(), name, 0, code, Parameter());
+        return Call::make(Handle(), Call::address_of, {ptr_0}, Call::Intrinsic);
+    }
+
+    Stmt call_extern_and_assert(const std::string& name, const std::vector<Expr>& args) {
+        Expr call = Call::make(Int(32), name, args, Call::Extern);
+        string call_result_name = unique_name(name + "_result");
+        Expr call_result_var = Variable::make(Int(32), call_result_name);
+        return LetStmt::make(call_result_name, call,
+                             AssertStmt::make(EQ::make(call_result_var, 0), call_result_var));
+    }
+
+    using IRMutator::visit;
+
+    void visit(const For *loop) {
+        if (loop->device_api == DeviceAPI::Hexagon) {
+            // Unrolling or loop partitioning might generate multiple
+            // loops with the same name, so we need to unique them.
+            std::string hex_name = "hex_" + loop->name;
+
+            // After moving this to Hexagon, it doesn't need to be
+            // marked Hexagon anymore.
+            Stmt body = For::make(loop->name, loop->min, loop->extent, loop->for_type,
+                                  DeviceAPI::None, loop->body);
+            body = remove_trivial_for_loops(body);
+
+            // Build a closure for the device code.
+            // TODO: Should this move the body of the loop to Hexagon,
+            // or the loop itself? Currently, this moves the loop itself.
+            Closure c(body);
+
+            // Make an argument list, and generate a function in the
+            // device_code module. The hexagon runtime code expects
+            // the arguments to appear in the order of (input buffers,
+            // output buffers, input scalars).  There's a huge hack
+            // here, in that the scalars must be last for the scalar
+            // arguments to shadow the symbols of the buffer.
+            std::vector<LoweredArgument> input_buffers, output_buffers;
+            std::map<std::string, Parameter> replacement_params;
+            for (const auto& i : c.buffers) {
+                if (i.second.write) {
+                    Argument::Kind kind = Argument::OutputBuffer;
+                    output_buffers.push_back(LoweredArgument(i.first, kind, i.second.type, i.second.dimensions));
+                } else {
+                    Argument::Kind kind = Argument::InputBuffer;
+                    input_buffers.push_back(LoweredArgument(i.first, kind, i.second.type, i.second.dimensions));
+                }
+
+                // Build a parameter to replace.
+                Parameter p(i.second.type, true, i.second.dimensions);
+                // Assert that buffers are aligned to one HVX
+                // vector. Generally, they should be page aligned ION
```

Comment:
I'm just going to remove that part of the comment, it's unnecessary here, and no one is going to attempt to call these hexagon functions directly (it would be a nightmare just to determine the arguments, let alone their meaning).